### ISSUE TITLE
Internal compiler error when building rust-sdl2, FulfillmentError

### ISSUE THREAD

**Issue by ca1ek** [rust-lang/rust#43947]
When attempting to build orbtk, building of it's dependency, sdl2, failed with an internal compiler error.

I tried this code: https://github.com/ca1ek/orbtk/blob/master/examples/list.rs

I expected to see this happen: The program being built and ran.

Instead, this happened: An internal compiler error.

## Meta

`rustc --version --verbose`:
```
rustc 1.20.0-nightly (15aa15b03 2017-07-21)
binary: rustc
commit-hash: 15aa15b03f41278aa5b9e56975889929ee734c63
commit-date: 2017-07-21
host: x86_64-unknown-linux-gnu
release: 1.20.0-nightly
LLVM version: 4.0
```

Backtrace:
```
   Compiling sdl2 v0.29.1
   Compiling image v0.12.4
error: internal compiler error: /checkout/src/librustc/infer/mod.rs:573: Encountered errors `[FulfillmentError(Obligation(predicate=Binder(TraitPredicate(<CB as audio::AudioCallback>)),depth=0),Unimplemented)]` resolving bounds after type-checking

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.20.0-nightly (15aa15b03 2017-07-21) running on x86_64-unknown-linux-gnu

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'Box<Any>', /checkout/src/librustc_errors/lib.rs:437:8
stack backtrace:
   0:     0x7fd51b18ba23 - std::sys::imp::backtrace::tracing::imp::unwind_backtrace::h84bea9eef7637c7f
                               at /checkout/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1:     0x7fd51b186104 - std::sys_common::backtrace::_print::h93d609d31b66fd57
                               at /checkout/src/libstd/sys_common/backtrace.rs:71
   2:     0x7fd51b1998a3 - std::panicking::default_hook::{{closure}}::h3da065dfd231ad18
                               at /checkout/src/libstd/sys_common/backtrace.rs:60
                               at /checkout/src/libstd/panicking.rs:380
   3:     0x7fd51b199598 - std::panicking::default_hook::he71f9097ed6368c5
                               at /checkout/src/libstd/panicking.rs:390
   4:     0x7fd51b199da7 - std::panicking::rust_panic_with_hook::hb7a0df8dd84ed7a6
                               at /checkout/src/libstd/panicking.rs:611
   5:     0x7fd517e4bf78 - std::panicking::begin_panic_new::h0925ce4a2f53555c
   6:     0x7fd517e36361 - rustc_errors::Handler::span_bug::h91bbf0b512c0181f
   7:     0x7fd518074623 - rustc::session::opt_span_bug_fmt::{{closure}}::h79da16154085c3e9
   8:     0x7fd518074342 - rustc::session::opt_span_bug_fmt::h019039aa36cacd98
   9:     0x7fd518074213 - rustc::session::span_bug_fmt::h36ead2eed7f132b1
  10:     0x7fd51812c59f - <&'gcx rustc::ty::TyS<'gcx> as rustc::infer::TransNormalize<'gcx>>::trans_normalize::hc514f88ca5b305f3
  11:     0x7fd517fee13d - rustc::infer::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'tcx>>::normalize_associated_type_in_env::hbb740490cbd5fce7
  12:     0x7fd5180ce721 - rustc::ty::layout::Layout::compute_uncached::hc58895a8da1a5c62
  13:     0x7fd5180f270a - rustc::ty::util::layout_raw::h47d964017e4a4426
  14:     0x7fd518199989 - rustc::ty::maps::<impl rustc::ty::maps::queries::layout_raw<'tcx>>::try_get::h00ac5eb525533633
  15:     0x7fd5181a676f - rustc::ty::maps::TyCtxtAt::layout_raw::hb425cd9eb9e1fdfa
  16:     0x7fd5181a187d - rustc::ty::maps::<impl rustc::ty::context::TyCtxt<'a, 'tcx, 'lcx>>::layout_raw::h58851a09d8125b7e
  17:     0x7fd518a0f404 - rustc_const_eval::eval::eval_const_expr_partial::{{closure}}::ha20b8c7519cbc3de
  18:     0x7fd518a095b3 - rustc_const_eval::eval::eval_const_expr_partial::h456c2783d8f411b4
  19:     0x7fd518a0927b - rustc_const_eval::eval::eval_const_expr_partial::h456c2783d8f411b4
  20:     0x7fd518a0927b - rustc_const_eval::eval::eval_const_expr_partial::h456c2783d8f411b4
  21:     0x7fd518a0cbf2 - rustc_const_eval::eval::eval_const_expr_partial::h456c2783d8f411b4
  22:     0x7fd518a08920 - rustc_const_eval::eval::ConstContext::eval::hdaf0068b2a6ccdd1
  23:     0x7fd51a8dc790 - <rustc_passes::consts::CheckCrateVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h28de0e95753a949e
  24:     0x7fd51a8dc625 - <rustc_passes::consts::CheckCrateVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h28de0e95753a949e
  25:     0x7fd51a8dc21e - <rustc_passes::consts::CheckCrateVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h28de0e95753a949e
  26:     0x7fd51a8dc417 - <rustc_passes::consts::CheckCrateVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h28de0e95753a949e
  27:     0x7fd51a8dc625 - <rustc_passes::consts::CheckCrateVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h28de0e95753a949e
  28:     0x7fd51a8db109 - <rustc_passes::consts::CheckCrateVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_nested_body::h5d9d2d59c7e9ec71
  29:     0x7fd51a8dce2c - rustc_passes::consts::check_crate::h0573abf01816161a
  30:     0x7fd51b53b395 - rustc_driver::driver::phase_3_run_analysis_passes::{{closure}}::hc32fabe122cd8e11
  31:     0x7fd51b5267a3 - rustc_driver::driver::phase_3_run_analysis_passes::h418664d006a1abd0
  32:     0x7fd51b51527d - rustc_driver::driver::compile_input::hfa49c353ee615e7e
  33:     0x7fd51b55a555 - rustc_driver::run_compiler::hc734749495546041
  34:     0x7fd51b4733d7 - std::sys_common::backtrace::__rust_begin_short_backtrace::h0094db030b02e908
  35:     0x7fd51b1c92ec - __rust_maybe_catch_panic
                               at /checkout/src/libpanic_unwind/lib.rs:98
  36:     0x7fd51b4aac00 - <F as alloc::boxed::FnBox<A>>::call_box::h2fec5da993954aab
  37:     0x7fd51b19879b - std::sys::imp::thread::Thread::new::thread_start::h33b55db0d5c58f41
                               at /checkout/src/liballoc/boxed.rs:661
                               at /checkout/src/libstd/sys_common/thread.rs:21
                               at /checkout/src/libstd/sys/unix/thread.rs:84
  38:     0x7fd51631d6d9 - start_thread
  39:     0x7fd51ae4317e - __clone
  40:                0x0 - <unknown>

error: Could not compile `sdl2`.
warning: build failed, waiting for other jobs to finish...
```

**Comment by kennytm** [comment]
You are using a very outdated nightly (1.20.0-nightly 2017-07-21 — that's a month ago). Please update the nightly and try again.

The ICE error looks exactly the same as #43391, and was closed as duplicate of #43394, which was closed as duplicate of #43357, which has been fixed in #43505 twenty days ago.

**Comment by ca1ek** [comment]
whoops. Didn't notice the date, sorry.
